hd44780/hd44780.go

```diff
@@ -3,176 +3,45 @@ package hd44780
+func NewGPIO8Bit(dataPins []uint8, e, rs, rw uint8) (Device, error) {
 	const eightBitMode = 8
-	if len(data) != eightBitMode {
+	if len(dataPins) != eightBitMode {
 		return Device{}, errors.New("8 pins are required in data slice (D7-D0) when HD44780 is used in 8 bit mode")
 	}
-	return newGPIO(data, e, rs, rw, DATA_LENGTH_8BIT), nil
-}
-
-func newGPIO(data []uint8, e, rs, rw uint8, mode byte) Device {
-	pins := make([]machine.GPIO, len(data))
-	for i := 0; i < len(data); i++ {
-		m := machine.GPIO{Pin: data[i]}
-		m.Configure(machine.GPIOConfig{Mode: machine.GPIO_OUTPUT})
-		pins[i] = m
-	}
-	enable := machine.GPIO{e}
-	enable.Configure(machine.GPIOConfig{Mode: machine.GPIO_OUTPUT})
-	registerSelect := machine.GPIO{rs}
-	registerSelect.Configure(machine.GPIOConfig{Mode: machine.GPIO_OUTPUT})
-	readWrite := machine.GPIO{rw}
-	readWrite.Configure(machine.GPIOConfig{Mode: machine.GPIO_OUTPUT})
-	readWrite.Low()
-
-	gpio := GPIO{
-		dataPins: pins,
-		e:        enable,
-		rs:       registerSelect,
-		rw:       readWrite,
-	}
-
-	if mode == DATA_LENGTH_4BIT {
-		gpio.write = gpio.write4BitMode
-		gpio.read = gpio.read4BitMode
-	} else {
-		gpio.write = gpio.write8BitMode
-		gpio.read = gpio.read8BitMode
-	}
-
-	return Device{
-		bus:        &gpio,
-		datalength: mode,
-	}
-}
-
-// SetCommandMode sets command/instruction mode
-func (g *GPIO) SetCommandMode(set bool) {
-	if set {
-		g.rs.Low()
-	} else {
-		g.rs.High()
-	}
-}
-
-func (g *GPIO) Write(data byte) {
-	g.rw.Low()
-	g.write(data)
-}
-
-func (g *GPIO) write8BitMode(data byte) {
-	g.e.High()
-	g.setPins(data)
-	g.e.Low()
-}
-
-func (g *GPIO) write4BitMode(data byte) {
-	g.e.High()
-	g.setPins(data >> 4)
-	g.e.Low()
-
-	g.e.High()
-	g.setPins(data)
-	g.e.Low()
-}
-
-func (g *GPIO) Read() byte {
-	g.rs.Low()
-	g.rw.High()
-	g.reconfigureGPIOMode(machine.GPIO_INPUT)
-	data := g.read()
-	g.reconfigureGPIOMode(machine.GPIO_OUTPUT)
-	return data
-}
-
-func (g *GPIO) read4BitMode() byte {
-	g.e.High()
-	data := (g.pins() << 4 & 0xF0)
-	g.e.Low()
-	g.e.High()
-	data |= (g.pins() & 0x0F)
-	g.e.Low()
-	return data
-}
-func (g *GPIO) read8BitMode() byte {
-	g.e.High()
-	data := g.pins()
-	g.e.Low()
-	return data
-}
-func (g *GPIO) reconfigureGPIOMode(mode machine.GPIOMode) {
-	for i := 0; i < len(g.dataPins); i++ {
-		g.dataPins[i].Configure(machine.GPIOConfig{Mode: mode})
-	}
-}
-
-// setPins sets high or low state on all data pins depending on data
-func (g *GPIO) setPins(data uint8) {
-	mask := uint8(1)
-	for i := 0; i < len(g.dataPins); i++ {
-		if (data & mask) != 0 {
-			g.dataPins[i].High()
-		} else {
-			g.dataPins[i].Low()
-		}
-		mask = mask << 1
-	}
-}
-
-// pins returns current state of data pins. MSB is D7
-func (g *GPIO) pins() uint8 {
-	bits := uint8(0)
-	for i := uint8(0); i < uint8(len(g.dataPins)); i++ {
-		if g.dataPins[i].Get() {
-			bits |= (1 << i)
-		}
-	}
-	return bits
+	return newGPIO(dataPins, e, rs, rw, DATA_LENGTH_8BIT), nil
 }
 
 type Device struct {
```

Comment:
I like to put all struct definitions on the beginning of the file, or at least before the functions that use it